src/index.ts

```diff
@@ -1,74 +1,196 @@
+    let baseVal = interpretNext(tailProto[0], tailProto[1] ?? '', restMa)
+    // We are reinterpreting the base as http, so we need to remove the tcp:// if it's there
+    baseVal = baseVal.replace('tcp://', '')
+    return `https://${baseVal}`
+  },
+  ws: (value: string, restMa: StringTuple[]) => {
+    const maHasTLS = hasTLS(restMa)
+    const sni = extractSNI(restMa)
+    if (maHasTLS && sni !== null) {
+      return `wss://${sni}`
+    }
+    const protocol = maHasTLS ? 'wss://' : 'ws://'
+    const tailProto = restMa.pop()
+    if (tailProto === undefined) {
+      throw new Error('Unexpected end of multiaddr')
+    }
+    let baseVal = interpretNext(tailProto[0], tailProto[1] ?? '', restMa)
+    // We are reinterpreting the base, so we need to remove the tcp:// if it's there
+    baseVal = baseVal.replace('tcp://', '')
+    return `${protocol}${baseVal}`
+  },
+  wss: (value: string, restMa: StringTuple[]) => {
+    const tailProto = restMa.pop()
+    if (tailProto === undefined) {
+      throw new Error('Unexpected end of multiaddr')
+    }
+    let baseVal = interpretNext(tailProto[0], tailProto[1] ?? '', restMa)
+    // We are reinterpreting the base as http, so we need to remove the tcp:// if it's there
+    baseVal = baseVal.replace('tcp://', '')
+    return `wss://${baseVal}`
+  },
+  'p2p-websocket-star': (value: string, restMa: StringTuple[]) => {
+    const tailProto = restMa.pop()
+    if (tailProto === undefined) {
+      throw new Error('Unexpected end of multiaddr')
+    }
+    return `${interpretNext(tailProto[0], tailProto[1] ?? '', restMa)}/p2p-websocket-star`
+  },
+  'p2p-webrtc-star': (value: string, restMa: StringTuple[]) => {
+    const tailProto = restMa.pop()
+    if (tailProto === undefined) {
+      throw new Error('Unexpected end of multiaddr')
+    }
+    return `${interpretNext(tailProto[0], tailProto[1] ?? '', restMa)}/p2p-webrtc-star`
+  },
+  'p2p-webrtc-direct': (value: string, restMa: StringTuple[]) => {
+    const tailProto = restMa.pop()
+    if (tailProto === undefined) {
+      throw new Error('Unexpected end of multiaddr')
+    }
+    return `${interpretNext(tailProto[0], tailProto[1] ?? '', restMa)}/p2p-webrtc-direct`
+  }
 }
 
 export function multiaddrToUri (input: Multiaddr | string | Uint8Array, opts?: MultiaddrToUriOpts): string {
   const ma = multiaddr(input)
-  const parts = ma.toString().split('/').slice(1)
-  return ma
-    .tuples()
-    .map(tuple => ({
-      protocol: parts.shift() ?? '',
-      content: (tuple[1] != null) ? parts.shift() ?? '' : ''
-    }))
-    .reduce((str: string, part: Part, i: number, parts: Part[]) => {
-      const reduce = Reducers[part.protocol]
-      if (reduce == null) {
-        throw new Error(`Unsupported protocol ${part.protocol}`)
+  const parts = ma.stringTuples()
+  const head = parts.pop()
+  if (head === undefined) {
+    throw new Error('Unexpected end of multiaddr')
+  }
+
+  let uri = interpreters[protocols(head[0]).name](head[1] ?? '', parts)
```

Comment:
Need to null-guard on `interpreters` here to avoid unhelpful error messages , something like:

```suggestion
  const protocol = protocols(head[0])
  const interpreter =  interpreters[protocol.name]
  
  if (interpreter == null) {
    throw new Error(`No interpreter found for ${protocol.name}`)
  }
  
  let uri = interpreter(head[1] ?? '', parts)
```